File: 0463-island-perimeter/0463-island-perimeter.py
class Solution:
    def islandPerimeter(self, grid: List[List[int]]) -> int:
        from collections import deque
        def bfs(tx, ty):
            dq = deque()
            dq.append((tx, ty))
            while dq:
                x, y = dq.popleft()
                if x < 0 or y < 0 or x >= len(grid) or y >= len(grid[0]):
                    continue
                if visited[x][y]:
                    continue
                visited[x][y] = 1
                #print("visited: ", visited)
                ans[0] += 4
                #print("ans: ", ans[0]) 
                for i in range(4):
                    nx, ny = x+dx[i], y+dy[i] 
                    if 0 <= nx < len(grid) and 0 <= ny < len(grid[0]):
                        if grid[nx][ny]:        # island
                            ans[0] -= 1
                            if not visited[nx][ny]:
                                dq.append([nx, ny])

        visited = [[0] * len(grid[0]) for _ in range(len(grid))]
        #print(visited)
        ans = [0]
        dx = [-1, 1, 0, 0]
        dy = [0, 0, -1, 1]

        flag = False
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if not visited[i][j]:
                    if grid[i][j]:
                        #print("call bfs: i, j : ", i, j)
                        bfs(i, j)
                        flag = True
                        break
            if flag:
                break

        return ans[0] 

File: 0463-island-perimeter/0463-island-perimeter.cpp
class Solution {
public:
    int islandPerimeter(vector<vector<int>>& grid) {
        vector<vector<int>> dir = {{0,1}, {-1,0}, {0,-1}, {1,0}};
        int m = grid.size(), n = grid[0].size();
        int ans = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == 0) continue;
                int temp = 4;
                for (int k = 0; k < 4; k++) {
                    int nx = i + dir[k][0];
                    int ny = j + dir[k][1];
                    if (0 <= nx && nx < m && 0 <= ny && ny < n && grid[nx][ny] == 1) temp--;
                }
                ans += temp;
            }
        } 
        return ans;
    }
};